domain/schema/secret.go

```diff
@@ -0,0 +1,334 @@
+        pending_delete BOOLEAN NOT NULL DEFAULT (FALSE),
+        create_time DATETIME NOT NULL DEFAULT (STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW', 'utc')),
+        update_time DATETIME NOT NULL DEFAULT (STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW', 'utc')),
+        CONSTRAINT fk_secret_revision_secret_uuid 
+            FOREIGN KEY (secret_uuid)
+            REFERENCES secret (uuid)
+    );
+
+CREATE UNIQUE INDEX idx_secret_revision_secret_uuid_revision ON secret_revision (secret_uuid,revision);
+
+CREATE TABLE
+    secret_revision_expire (
+        revision_uuid TEXT PRIMARY KEY,
+        next_expire_time DATETIME NOT NULL,
+        CONSTRAINT fk_secret_revision_expire_revision_uuid
+            FOREIGN KEY (revision_uuid)
+            REFERENCES secret_revision (uuid)
+    );
+
+CREATE TABLE
+    secret_application_owner (
+        secret_uuid TEXT PRIMARY KEY,
+        application_uuid TEXT NOT NULL,
+        label TEXT,
+        CONSTRAINT fk_secret_application_owner_secret_uuid
+            FOREIGN KEY (secret_uuid)
+            REFERENCES secret (uuid),
+        CONSTRAINT fk_secret_application_owner_application_uuid
+            FOREIGN KEY (application_uuid)
+            REFERENCES application (uuid)
+    );
+
+-- We need to ensure the label is unique per the application.
+CREATE UNIQUE INDEX idx_secret_application_owner_label ON secret_application_owner (label,application_uuid);
+
+CREATE TABLE
+    secret_unit_owner (
+        secret_uuid TEXT PRIMARY KEY,
+        unit_uuid TEXT NOT NULL,
+        label TEXT,
+        CONSTRAINT fk_secret_unit_owner_secret_uuid
+            FOREIGN KEY (secret_uuid)
+            REFERENCES secret (uuid),
+        CONSTRAINT fk_secret_unit_owner_unit_uuid
+            FOREIGN KEY (unit_uuid)
+            REFERENCES unit (uuid)
+    );
+
+-- We need to ensure the label is unique per unit.
+CREATE UNIQUE INDEX idx_secret_unit_owner_label ON secret_unit_owner (label,unit_uuid);
+
+CREATE TABLE
+    secret_model_owner (
+        secret_uuid TEXT PRIMARY KEY,
+        label TEXT,
+        CONSTRAINT fk_secret_model_owner_secret_uuid
+            FOREIGN KEY (secret_uuid)
+            REFERENCES secret (uuid)
+    );
+
+CREATE UNIQUE INDEX idx_secret_model_owner_label ON secret_model_owner (label);
+
+CREATE TABLE
+    secret_application_consumer (
+        secret_uuid TEXT NOT NULL,
+        application_uuid TEXT NOT NULL,
+        label TEXT,
+        current_revision INT NOT NULL,
+        CONSTRAINT pk_secret_application_consumer_secret_uuid_application_uuid
+            PRIMARY KEY (secret_uuid, application_uuid),
+        CONSTRAINT fk_secret_application_consumer_secret_uuid
+            FOREIGN KEY (secret_uuid)
+            REFERENCES secret (uuid),
+        CONSTRAINT fk_secret_application_consumer_application_uuid
+            FOREIGN KEY (application_uuid)
+            REFERENCES application (uuid)
+    );
+
+CREATE TRIGGER secret_application_consumer_delete
```

Comment:
Can this deletion happen outside of our domain invocations? If not, I'd probably prefer to have this DML in the state as an explicit action accompanying consumer deletion rather than in a trigger. That way everything that happens is apparent to a reader of that code.